### PR TITLE
MsgPack bin8/bin16/bin32 support

### LINKED ISSUE #922: Add binary support for MessagePack (C4, C5, C6)

### PR DESCRIPTION
**Pull request by Sanae6**
I've added binary support to the msgpack deserializer and serializer, along with a fallback for the json serializer. I've also added some tests to the relevant files. Let me know if there's anything I can do to make this implementation better. Thank you for the amazing library :)

This would fix and close #922.

### CODE REVIEW

**Comment by bblanchon** [comment]
Hi @Sanae6,

Thank you for this excellent PR!

Before starting the review, I'd like you to remove `LinkedBinary`.
Indeed, binary values should be treated like raw JSON strings and always be copied; this is part of a larger (incomplete) refactoring to reduce the size of `VariantContent`.
Then, of course, simplify the names accordingly, since we no longer need the distinction between "owned" and "linked" binary values.

Best regards,
Benoit

**Comment by Sanae6** [comment]
I removed LinkedBinaryValue and renamed OwnedBinaryValue to BinaryValue. Not sure what to name [Binary](https://github.com/bblanchon/ArduinoJson/blob/110e63981c4871217d42981b31aaa76abf1b597e/src/ArduinoJson/MsgPack/BinaryValue.hpp#L79), but I left it as something sensible enough.

File: src/ArduinoJson/MsgPack/BinaryValue.hpp [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+  }
+
+ private:
+  detail::StringNode* stringNode_;
+};
+
+using Binary = BinaryValue<detail::StringNode*>;
+
+template <typename T>
+inline BinaryValue<T> binary(const T& value) {
```
Comment (bblanchon) [inline src/ArduinoJson/MsgPack/BinaryValue.hpp]: The word `binary` is too generic.
Please rename the class to `MsgPackBinary` and remove the function.

File: src/ArduinoJson/MsgPack/BinaryValue.hpp [hunk 2]
```diff
@@ -0,0 +1,91 @@
+#pragma once
+
+ARDUINOJSON_BEGIN_PUBLIC_NAMESPACE
+
+template <typename T>
+class BinaryValue {
```
Comment (bblanchon) [inline src/ArduinoJson/MsgPack/BinaryValue.hpp]: Supporting container is cool, but not worth the complexity.
Please simplify the class to support only `const void*` / `size_t`.

File: src/ArduinoJson/MsgPack/BinaryValue.hpp [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    return size_ * sizeof(T);
+  }
+
+ private:
+  const T* data_;
+  size_t size_;
+};
+
+template <>
+class BinaryValue<detail::StringNode*> {
```
Comment (bblanchon) [inline src/ArduinoJson/MsgPack/BinaryValue.hpp]: I'm surprised to see this.
Please remove this specialization.

File: extras/tests/MsgPackDeserializer/deserializeVariant.cpp [hunk 4]
```diff
@@ -139,6 +140,55 @@ TEST_CASE("deserialize MsgPack value") {
...
+
+    REQUIRE(error == DeserializationError::Ok);
+    REQUIRE(doc.is<Binary>());
+    auto binary = doc.as<Binary>();
+    REQUIRE(binary.size_bytes() == 0x100);
+    REQUIRE(binary.data() != nullptr);
+    REQUIRE(reinterpret_cast<const char*>(binary.data())[0] == 5);
+  }
+
+#if ARDUINOJSON_STRING_LENGTH_SIZE >= 4
```
Comment (bblanchon) [inline extras/tests/MsgPackDeserializer/deserializeVariant.cpp]: Please move this test to a dedicated file in `MixedConfiguration`, so you can control the value of `ARDUINOJSON_STRING_LENGTH_SIZE`

File: extras/tests/MsgPackDeserializer/errors.cpp [hunk 5]
```diff
@@ -134,18 +134,6 @@ TEST_CASE("deserializeMsgPack() returns IncompleteInput") {
...
-  }
-
-  SECTION("bin 16") {
-    testIncompleteInput("\xc5\x00\x01X", 4);
-  }
-
-  SECTION("bin 32") {
-    testIncompleteInput("\xc6\x00\x00\x00\x01X", 6);
-  }
-
```
Comment (bblanchon) [inline extras/tests/MsgPackDeserializer/errors.cpp]: These tests should still pass.
Please restore them.

File: extras/tests/MsgPackDeserializer/filter.cpp [hunk 6]
```diff
@@ -375,52 +375,6 @@ TEST_CASE("deserializeMsgPack() filter") {
...
-        CHECK(doc.as<std::string>() == "{\"include\":42}");
-        CHECK(spy.log() ==
-              AllocatorLog{
-                  Allocate(sizeofStringBuffer()),
-                  Reallocate(sizeofStringBuffer(), sizeofString("include")),
-                  Allocate(sizeofPool()),
-                  Reallocate(sizeofPool(), sizeofObject(1)),
-              });
-      }
-
```
Comment (bblanchon) [inline extras/tests/MsgPackDeserializer/filter.cpp]: These tests should still pass.
Please restore them.

File: src/ArduinoJson/Configuration.hpp [hunk 7]
```diff
@@ -134,7 +134,7 @@
 #  if ARDUINOJSON_SIZEOF_POINTER <= 2
 #    define ARDUINOJSON_STRING_LENGTH_SIZE 1  // up to 255 characters
 #  else
-#    define ARDUINOJSON_STRING_LENGTH_SIZE 2  // up to 65535 characters
+#    define ARDUINOJSON_STRING_LENGTH_SIZE 4  // up to 65535 characters
```
Comment (bblanchon) [inline src/ArduinoJson/Configuration.hpp]: Please revert this change.
You can change settings in the compilation units of the `MixedConfiguration` executable.

File: src/ArduinoJson/Json/JsonSerializer.hpp
Comment (bblanchon) [inline src/ArduinoJson/Json/JsonSerializer.hpp:1]: This feature is for message pack only; other users should not be affected.
Besides, JSON doesn't support hexadecimal.
Please revert all changes affecting the JSON related code.

File: src/ArduinoJson/Variant/VariantContent.hpp [hunk 8]
```diff
@@ -31,7 +31,8 @@ enum {
   VALUE_IS_SIGNED_INTEGER = 0x0A,
   VALUE_IS_FLOAT = 0x0C,
 
-  COLLECTION_MASK = 0x60,
+  COLLECTION_MASK = 0x70,
```
Comment (bblanchon) [inline src/ArduinoJson/Variant/VariantContent.hpp]: Collections in this context are arrays and objects, but not binary values.
Please restore the original value.

**Review by bblanchon — CHANGES_REQUESTED** [review]

File: src/ArduinoJson/MsgPack/BinaryValue.hpp [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+
+  const T* begin() const {
+    return data_;
+  }
+
+  const T* end() const {
+    return data_ + size_;
+  }
+
+  size_t size_bytes() const {
```
Comment (bblanchon) [inline src/ArduinoJson/MsgPack/BinaryValue.hpp]: Please rename `size_bytes()` to `size` to be consistent with the rest of the library.

**Review by bblanchon — CHANGES_REQUESTED** [review]

File: src/ArduinoJson/Json/JsonSerializer.hpp
Comment (bblanchon) [reply src/ArduinoJson/Json/JsonSerializer.hpp:1]: Since you have to implement `visitBinary()`, please emit `null`.
I'll tell you later how I plan to eliminate this.

File: src/ArduinoJson/Variant/VariantContent.hpp (hunk 8 above)
Comment (Sanae6) [reply src/ArduinoJson/Variant/VariantContent.hpp]: Should I move `VALUE_IS_BINARY` out of the collections section? If so, where?

Comment (bblanchon) [reply src/ArduinoJson/Variant/VariantContent.hpp]: You should use the value 0x02 which used to be reserved for linked raw strings.

Comment (Sanae6) [reply src/ArduinoJson/Variant/VariantContent.hpp]: That doesn't seem right, that would make it so that binaries aren't owned by the document, which isn't true. I have it on 0x07 locally to make sure the owned flag is set right now. Is 0x02 really fine?

Comment (bblanchon) [reply src/ArduinoJson/Variant/VariantContent.hpp]: You're right. It doesn't work. Sorry.

I think I oversaw this problem because my plan was to remove `VALUE_IS_BINARY` and handle binary strings with `VALUE_IS_RAW_STRING` and let the converters do all the work.
This way, we add support for MessagePack binary strings without affecting JSON users and also pave the way to support MessagePack ext values in the future.

Nevertheless, I think it's still a good stepping stone to have the `VALUE_IS_BINARY`, even if it'll soon be removed.
Please ignore my previous comment about the value, and position the enum value where it makes sense to you. Don't overthink it, since we may remove it in a future commit.

File: extras/tests/MixedConfiguration/use_string_length_size_4.cpp
Comment (Sanae6) [inline extras/tests/MixedConfiguration/use_string_length_size_4.cpp:1]: This test currently fails. It seems that even when I apply the string length size, it doesn't actually apply to `StringNode::length_type`. Printing `StringNode::maxLength` in the `StringNode::create` function returned `0xffff` instead of `0xffffffff`. Any advice on how to solve that?

Comment (bblanchon) [reply extras/tests/MixedConfiguration/use_string_length_size_4.cpp:1]: This is surely due to the fact that the namespace doesn't include the value of `ARDUINOJSON_STRING_LENGTH_SIZE`.
**Let me fix that.**

Each file in this folder is named according to the setting name, so this file should be named `string_length_size_4.cpp`.

Anyway, thank you again for this excellent work 👍

File: extras/tests/JsonSerializer/JsonVariant.cpp [hunk 10]
```diff
@@ -3,6 +3,7 @@
 // MIT License
 
 #include <ArduinoJson.h>
+#include <array>
```
Comment (bblanchon) [inline extras/tests/JsonSerializer/JsonVariant.cpp]: This isn't needed anymore.

File: src/ArduinoJson/Json/JsonSerializer.hpp [hunk 11]
```diff
@@ -81,6 +81,11 @@ class JsonSerializer : public VariantDataVisitor<size_t> {
     return bytesWritten();
   }
 
+  size_t visit(MsgPackBinary) {
+    formatter_.writeRaw("null");
```
Comment (bblanchon) [inline src/ArduinoJson/Json/JsonSerializer.hpp]: I suspect that calling `visit(nullptr)` would lead to a smaller code.

File: src/ArduinoJson/Memory/StringNode.hpp [hunk 12]
```diff
@@ -11,6 +11,7 @@
 #include <ArduinoJson/Polyfills/limits.hpp>
 
 #include <stddef.h>  // offsetof
+#include <stdio.h>
```
Comment (bblanchon) [inline src/ArduinoJson/Memory/StringNode.hpp]: This was probably added by accident.
Please remove it.

**Review by bblanchon — CHANGES_REQUESTED** [review]

File: extras/tests/MixedConfiguration/use_string_length_size_4.cpp
Comment (Sanae6) [reply extras/tests/MixedConfiguration/use_string_length_size_4.cpp:1]: Alright! I'll leave it to you :)

**Review by bblanchon — CHANGES_REQUESTED** [review]
I just pushed the changes to fix this issue with `string_length_size_4.cpp`.
Please rebase your branch onto `7.x`.
Note that I created `string_length_size_{1,2,4}.cpp` to check that the fix actually works, so you'll have a conflict on `string_length_size_4.cpp`.

**Comment by Sanae6** [comment]
Alright! Force pushed, seems nothing is broken :+1:

File: extras/tests/MixedConfiguration/CMakeLists.txt [hunk 13]
```diff
@@ -22,6 +22,7 @@ add_executable(MixedConfigurationTests
 	use_double_1.cpp
 	use_long_long_0.cpp
 	use_long_long_1.cpp
+	string_length_size_4.cpp
```
Comment (bblanchon) [inline extras/tests/MixedConfiguration/CMakeLists.txt]: You can remove this line as `string_length_size_4.cpp` is already included.

**Comment by bblanchon** [comment]
It seems that you lost your version of `string_length_size_4.cpp` while merging the conflict.

**Review by bblanchon — CHANGES_REQUESTED** [review]
This looks really good.
I'll do an in-depth review tomorrow.
Thank you for this outstanding PR.

File: extras/tests/MsgPackSerializer/serializeVariant.cpp [hunk 14]
```diff
@@ -146,6 +147,19 @@ TEST_CASE("serialize MsgPack value") {
     checkVariant(serialized("\xDB\x00\x01\x00\x00", 5), "\xDB\x00\x01\x00\x00");
   }
 
+  SECTION("bin 8") {
+    static const auto bin8 = std::array<char, 4>({1, 2, 3, 4});
```
Comment (bblanchon) [inline extras/tests/MsgPackSerializer/serializeVariant.cpp]: `static const`?

File: extras/tests/MsgPackSerializer/serializeVariant.cpp [hunk 15]
```diff
@@ -146,6 +147,19 @@ TEST_CASE("serialize MsgPack value") {
...
+    checkVariant(MsgPackBinary(bin8.data(), bin8.size()),
+                 "\xC4\x04\x01\x02\x03\x04");
+  }
+
+  SECTION("bin 16") {
+    static const auto bin16 = std::array<char, 0x100>({1, 2, 3, 4});
```
Comment (bblanchon) [inline extras/tests/MsgPackSerializer/serializeVariant.cpp]: `static const`?

File: extras/tests/MsgPackDeserializer/deserializeVariant.cpp
Comment (bblanchon) [inline extras/tests/MsgPackDeserializer/deserializeVariant.cpp:1]: Remember to restore the `bin 32` test that you originally wrote in `string_length_size_4.cpp` but lost after rebasing.

File: extras/tests/MsgPackSerializer/serializeVariant.cpp [hunk 16]
```diff
@@ -146,6 +147,19 @@ TEST_CASE("serialize MsgPack value") {
...
+    std::string bin16string(bin16.data(), bin16.size());
```
Comment (bblanchon) [inline extras/tests/MsgPackSerializer/serializeVariant.cpp]: You could simply use `std::string(0x100, '?')`; this constructor creates a string that repeats the provided character N times.
See https://en.cppreference.com/w/cpp/string/basic_string/basic_string (2)

File: src/ArduinoJson/MsgPack/MsgPackBinary.hpp [hunk 17]
```diff
@@ -0,0 +1,25 @@
+#pragma once
+
+ARDUINOJSON_BEGIN_PUBLIC_NAMESPACE
+
+class MsgPackBinary {
+ public:
+  explicit MsgPackBinary(const void* c, size_t size) : data_(c), size_(size) {}
+  operator const void*() const {
```
Comment (bblanchon) [inline src/ArduinoJson/MsgPack/MsgPackBinary.hpp]: I don't think the implicit conversion to `const void*` is a good idea.
Please remove it.

File: src/ArduinoJson/Variant/VariantCompare.hpp [hunk 18]
```diff
@@ -164,6 +164,12 @@ struct VariantComparer : ComparerBase {
     return reverseResult(comparer);
   }
 
+  CompareResult visit(MsgPackBinary value) {
```
Comment (bblanchon) [inline src/ArduinoJson/Variant/VariantCompare.hpp:186]: I doubt this is covered by any test.

**Review by bblanchon — CHANGES_REQUESTED** [review]

**Comment by Sanae6** [comment]
Made the changes you wanted, testing MsgPackBinary comparison actually revealed my initial implementation didn't work, so I fixed that as well.

File: src/ArduinoJson/MsgPack/MsgPackSerializer.hpp [hunk 19]
```diff
@@ -123,6 +123,21 @@ class MsgPackSerializer : public VariantDataVisitor<size_t> {
...
 
+  size_t visit(MsgPackBinary value) {
+    if (value.size() <= 0xFF) {
+      writeByte(0xC4);
+      writeInteger(uint8_t(value.size()));
+    } else if (value.size() <= 0xFFFF) {
+      writeByte(0xC5);
+      writeInteger(uint16_t(value.size()));
+    } else {
+      writeByte(0xC6);
```
Comment (bblanchon) [inline src/ArduinoJson/MsgPack/MsgPackSerializer.hpp:134]: This line is not covered by the tests.
Please add a test in `string_length_size_4.cpp`

File: src/ArduinoJson/Variant/VariantData.hpp [hunk 20]
```diff
@@ -185,6 +190,16 @@ class VariantData {
     }
   }
 
+  MsgPackBinary asBinary() const {
+    switch (type()) {
+      case VALUE_IS_BINARY:
+        return MsgPackBinary(content_.asOwnedString->data,
+                             content_.asOwnedString->length);
+      default:
+        return MsgPackBinary(nullptr, 0);
```
Comment (bblanchon) [inline src/ArduinoJson/Variant/VariantData.hpp:199]: This line is not covered by the tests.

File: src/ArduinoJson/Variant/VariantData.hpp [hunk 21]
```diff
@@ -394,6 +413,28 @@ class VariantData {
...
+    if (dup)
+      setBinary(dup);
+    else
+      setNull();
+  }
+
+  static void setBinary(VariantData* var, MsgPackBinary value,
+                        ResourceManager* resources) {
+    if (!var)
+      return;
```
Comment (bblanchon) [inline src/ArduinoJson/Variant/VariantData.hpp:434]: This line is not covered by the tests.

**Review by bblanchon — CHANGES_REQUESTED** [review]

**Comment by bblanchon** [comment]
I added the three missing tests, squashed, and merged.

Thank you again for this excellent contribution ❤️

**Comment by bblanchon** [comment]
This feature is available in [ArduinoJson 7.1.0](https://github.com/bblanchon/ArduinoJson/releases/tag/v7.1.0)
